### ISSUE TITLE
macOS screen capture freezes randomly

### ISSUE THREAD

**Issue by BartlomiejLewandowski** [obsproject/obs-studio#11435]
### Operating System Info

macOS 14

### OBS Studio Version

30.2.3

### OBS Studio Log URL

https://obsproject.com/logs/TOdgwuvLpRtJSDOQ

### Expected Behavior

Recording and screen capture working without issues

### Current Behavior

Randomly, but happens quite often, the macOSScreenCap source no longer updates when the screen is being updated. 
The workaround is to open the properties of the source and click on Restart Capture. 

The UI is updated to show the latest capture and everything works smoothly. Until it happens again.

### Steps to Reproduce

1. Add source - screen capture on a mac
2. Wait some time
3. The capture no longer updates when the screen changes

### Anything else we should know?

Interestingly after clicking the button (restart capture) is disabled. Is it only enabled when we detect something is wrong?
I'd like to understand where the issue is.

**Comment by jcm93** [comment]
This has been kind of a known issue since SCK has existed. It used to be prevalent mostly under memory pressure, and so lower memory configurations would run into it more often. It is also known to occur if you have a relatively low amount of free disk space (less free disk space than RAM, for example).

Outside of that, it does seem to still happen semi-randomly after prolonged periods (several hours of recording or streaming).

Important to note that sources freezing if the machine goes to sleep is both known and expected; should be sure that this is not the case on your machine.

Possible dupe of: https://github.com/obsproject/obs-studio/issues/8320, https://github.com/obsproject/obs-studio/issues/9056

**Comment by BartlomiejLewandowski** [comment]
Yes, I'm recording a workshop, so I'm on the machine the whole time. Can you shed some light on the code behind disabling the Restart Capture button? When does the state of that button change?

**Comment by jcm93** [comment]
Restart Capture is only activated in cases where OBS explicitly receives a "the stream stopped" message from ScreenCaptureKit (the underlying API); i.e. SCK reported an error, or the user pressed Apple's little extra menu bar control to stop the capture.

In the cases where captures freeze inexplicably, most frequently the cause is `replayd` crashing (though not always). When `replayd` crashes, no error reaches OBS and so OBS is not aware that the capture has stopped; it pretty much just stops receiving frames.

**Comment by BartlomiejLewandowski** [comment]
Can OBS detect that no new frames are being received while it expect those to arrive? It sounds like a more robust solution, instead of reacting on a error that might never arrive.

**Comment by jcm93** [comment]
There are many situations in normal operation where you won't receive a new frame for a prolonged period of time. SCK only delivers frames when there are actual updates, to remain efficient. Differentiating between a situation where content is temporarily static versus one where a stream has frozen would be very nontrivial.

One approach might be to try and closely monitor `replayd` to see if it crashes, but there isn't a great way to do that, and moreover, that shouldn't really be OBS's job; the system should just be more reliable.

I would confirm that you have ample free disk space, more than you ought to need, and see if that alleviates the issue for you. I'd also file feedback with Apple, since the issue here ultimately lies with them.

**Comment by BartlomiejLewandowski** [comment]
I agree, wasn't aware that the stream doesn't contain some sort of a keepalive. I'll file feedback then. Cheers!
